src/scene/light.js

```diff
@@ -311,6 +320,112 @@ Object.assign(Light.prototype, {
         }
 
         this.key = key;
+    },
+
+    // creates LUT texture used by area lights
+    uploadAreaLightLUTs: function () {
+
+        function createTexture(device, data, format) {
+            var tex = new pc.Texture(device, {
+                width: 64,
+                height: 64,
+                format: format,
+                addressU: pc.ADDRESS_CLAMP_TO_EDGE,
+                addressV: pc.ADDRESS_CLAMP_TO_EDGE,
+                type: pc.TEXTURETYPE_RGBE,
+                magFilter: pc.FILTER_LINEAR,
+                minFilter: pc.FILTER_NEAREST,
+                anisotropy: 1
+            });
+
+            tex.lock().set(data);
+            tex.unlock();
+            tex.upload();
+
+            return tex;
+        }
+
+        function offsetScale(data, offset, scale) {
+
+            var count = data.length;
+            var ret = new Float32Array(count);
+            for (var i = 0; i < count; i++) {
+                var n = i % 4;
+                ret[i] = (data[i] + offset[n]) * scale[n];
+            }
+            return ret;
+        }
+
+        function convertToHalfFloat(data) {
+
+            var count = data.length;
+            var ret = new Uint16Array(count);
+            var float2Half = math.float2Half;
+            for (var i = 0; i < count; i++) {
+                ret[i] = float2Half(data[i]);
+            }
+
+            return ret;
+        }
+
+        function convertToUint(data){
+
+            var count = data.length;
+            var ret = new Uint8ClampedArray(count);
+            for (var i = 0; i < count; i++) {
+                ret[i] = data[i] * 255;
+            }
+
+            return ret;
+        }
+
+        // create texture if not already
+        var app = Application.getApplication();
```

Comment:
The scene layer should not have a dependency on the `Application`. This forms a circular dependency.

Reply:
this is existing pattern in this file already, as constructor does not take device and so this is how its obtained. I think this needs more global solution as it's taking place in few places.

Reply:
OK. I just think we need to resolve this soon because otherwise, we'll simply continue in the wrong direction.

Reply:
created: https://github.com/playcanvas/engine/issues/2651